fastplotlib/layouts/_gridplot.py

```diff
@@ -25,128 +25,197 @@ def to_array(a) -> np.ndarray:
+                controller_ids = ids_init
 
-        if cameras.shape != self.shape:
-            raise ValueError
-
-        # create controllers if the arguments were integers
-        if np.issubdtype(controllers.dtype, np.integer):
-            if not np.all(
-                np.sort(np.unique(controllers))
-                == np.arange(np.unique(controllers).size)
-            ):
-                raise ValueError("controllers must be consecutive integers")
-
-            for controller in np.unique(controllers):
-                cam = np.unique(cameras[controllers == controller])
-                if cam.size > 1:
-                    raise ValueError(
-                        f"Controller id: {controller} has been assigned to multiple different camera types"
-                    )
+            # integer ids
+            elif all([isinstance(item, (int, np.integer)) for item in flat]):
+                controller_ids = to_array(controller_ids).reshape(self.shape)
 
-                self._controllers[controllers == controller] = create_controller(cam[0])
-        # else assume it's a single pygfx.Controller instance or a list of controllers
-        else:
-            if isinstance(controllers, pygfx.Controller):
-                self._controllers = np.array(
-                    [controllers] * shape[0] * shape[1]
-                ).reshape(shape)
             else:
-                self._controllers = np.array(controllers).reshape(shape)
+                raise TypeError(
+                    f"list argument to `controller_ids` must be a list of `str` or `int`, "
+                    f"you have passed: {controller_ids}"
+                )
+
+        if controller_ids.shape != self.shape:
+            raise ValueError("Number of controller_ids does not match the number of subplots")
+
+        if controller_types is None:
+            # `create_controller()` will auto-determine controller for each subplot based on defaults
+            controller_types = np.array(["default"] * self.shape[0] * self.shape[1]).reshape(self.shape)
+
+        # validate controller types
+        flat = list(chain(*controller_types))
+        # str controller_type or pygfx instances
+        valid_str = list(valid_controller_types.keys()) + ["default"]
+        valid_instances = tuple(valid_controller_types.values())
+
+        # make sure each controller type is valid
+        for controller_type in flat:
+            if controller_type is None:
+                continue
+
+            if (controller_type not in valid_str) and (not isinstance(controller_type, valid_instances)):
+                raise ValueError(
+                    f"You have passed an invalid controller type, valid controller_types arguments are:\n"
+                    f"{valid_str} or instances of {[c.__name__ for c in valid_instances]}"
+                )
+
+        controller_types = to_array(controller_types).reshape(self.shape)
```

Comment:
should this be user-facing as read only?

Reply:
Similar to discussion above, right now the subplots own the controllers.